gremlin-go/driver/connection.go

```diff
@@ -19,50 +19,64 @@ under the License.
 
 package gremlingo
 
+import (
+	"errors"
+)
+
+type connectionState int
+
+const (
+	initialized connectionState = iota + 1
+	established
+	closed
+	closedDueToError
+)
+
 type connection struct {
-	host            string
-	port            int
-	transporterType TransporterType
-	logHandler      *logHandler
-	transporter     transporter
-	protocol        protocol
-	results         map[string]ResultSet
+	logHandler *logHandler
+	protocol   protocol
+	results    map[string]ResultSet
+	state      connectionState
 }
 
-func (connection *connection) close() (err error) {
-	if connection.transporter != nil {
-		err = connection.transporter.Close()
-	}
-	return
+func (connection *connection) errorCallback() {
+	connection.logHandler.log(Error, errorCallback)
+	connection.state = closedDueToError
+	_ = connection.protocol.close()
 }
 
-func (connection *connection) connect() error {
-	if connection.transporter != nil {
-		closeErr := connection.transporter.Close()
-		connection.logHandler.logf(Warning, transportCloseFailed, closeErr)
+func (connection *connection) close() error {
+	if connection.state != established {
+		return errors.New("cannot close connection that has already been closed or has not been connected")
 	}
-	connection.protocol = newGremlinServerWSProtocol(connection.logHandler)
-	connection.transporter = getTransportLayer(connection.transporterType, connection.host, connection.port)
-	err := connection.transporter.Connect()
-	if err != nil {
-		return err
+	connection.logHandler.log(Info, closeConnection)
+	var err error
+	if connection.protocol != nil {
+		err = connection.protocol.close()
 	}
-	connection.protocol.connectionMade(connection.transporter)
-	return nil
+	connection.state = closed
```

Comment:
here and everywhere else `connection` is accessed/mutated: does `connection` need to be safe to access concurrently? If so, we might want to introduce a mutex for locking